### PR TITLE
Fix missing transaction in light mempool bug

### PR DESCRIPTION
**Pull request by Shourya742**
Closes: #840

### CODE REVIEW

**Comment by GitGab19** [comment]
@Shourya742 I had a quick look at this solution and I have two major points to discuss:
- You remove all the txs contained in an old job from the mempool --> probably the majority of the txs contained in the new job will be the same --> it doesn't make too much sense to remove and add them again 
- To properly remove old txs, you transform the mempool from txid to short_id (which doesn't scale well as approach and it's already a bottleneck we encounter when we receive a `DeclareMiningJob` message.

Have you explored the idea about not removing the old txs, but simply updating the mempool (as it's already done) and adding the txs contained in the last declared jobs (if they are not already there in the mempool after the update)? 
Does it make sense to you? If not, which could be the issues?

**Comment by Shourya742** [comment]
> @Shourya742 I had a quick look at this solution and I have two major points to discuss:
> 
> * You remove all the txs contained in an old job from the mempool --> probably the majority of the txs contained in the new job will be the same --> it doesn't make too much sense to remove and add them again
> * To properly remove old txs, you transform the mempool from `txid` to `short_id` (which doesn't scale well as approach and it's already a bottleneck we encounter when we receive a `DeclareMiningJob` message.
> 
> Have you explored the idea about not removing the old txs, but simply updating the mempool (as it's already done) and adding the txs contained in the last declared jobs (if they are not already there in the mempool after the update)? Does it make sense to you? If not, which could be the issues?

Here's a refined and grammatically improved version of your message:

---
**Point 1**: We are checking if the `txid` is not present in the new job. Based on this, we decrement the counter value rather than deleting everything from the old job.

**Point 2**: I didn’t know that—let me check on it.

I initially thought keeping the JDS mempool as lean as possible was a requirement. However, it does make sense to let transactions persist in the mempool without a specific logic to remove them once a new DMJ is anticipated. In my testing, while we have logic to remove transactions from the old job, it didn’t result in a significant reduction in the number of transactions deleted from the mempool. But if we’re okay with not keeping the JDS mempool lean, then we can follow the approach you suggested.

**Review by GitGab19 — COMMENTED** [review]
LGTM.
I would change the `[mempool_update_interval]` parameter to 0.1 secs as we tested together.

File: roles/jd-server/src/lib/job_declarator/mod.rs [hunk 1]
```diff
@@ -145,7 +145,7 @@ impl JobDeclaratorDownstream {
                     .ok_or(Box::new(JdsError::ImpossibleToReconstructBlock(
                         "Txid found in jds mempool but transactions not present".to_string(),
                     )))?;
-                transactions_list.push(tx);
+                transactions_list.push(tx.0);
```
Comment (GitGab19) [inline roles/jd-server/src/lib/job_declarator/mod.rs:148]: Why did you change this?

**Review by GitGab19 — COMMENTED** [review]
LGTM. I just left a minor comment. I would say that we could add an integration test for this scenario in the future. I remember we were tracking these issues to be tested by the new integration tests framework somewhere, but I can't find it anymore (cc @plebhash)

File: roles/jd-server/src/lib/job_declarator/mod.rs (hunk 1 above)
Comment (Shourya742) [reply roles/jd-server/src/lib/job_declarator/mod.rs:148]: Since the hashmap structure is now `(txid, (tx, count))` instead of `(txid, tx)`, we used the first element of the inner tuple to reference `tx`.

**Review by GitGab19 — APPROVED** [review]

**Comment by plebhash** [comment]
> LGTM. I just left a minor comment. I would say that we could add an integration test for this scenario in the future. I remember we were tracking these issues to be tested by the new integration tests framework somewhere, but I can't find it anymore (cc @plebhash)

it's https://github.com/stratum-mining/stratum/issues/1077
